main/lsp/notifications/sorbet_workspace_edit.cc

```diff
@@ -87,6 +94,17 @@ void SorbetWorkspaceEditTask::runSpecial(LSPTypechecker &typechecker, WorkerPool
     if (latencyTimer != nullptr) {
         latencyTimer->setTag("path", "slow");
     }
+    if (indexer) {
+        ENFORCE(updates == nullptr);
+        // Using the `indexer` here is safe; the indexing thread is blocked until `startedNotification.Notify` is called
+        // later in this function.
+        // This is really gnarly; there's got to be a cleaner way to do threading here. We can't move this out because
+        // we need `workers`, which is a resource that is explicitly managed by typechecking.
+        updates = make_unique<LSPFileUpdates>(indexer->commitEdit(*params, workers));
```

Comment:
Generally speaking, LSP has an indexer thread that runs `commitEdit` (to maintain a GlobalState with a name table that we use as the 'seed' for the slow path) and a typechecking thread that actually typechecks files. The typechecking thread owns and manages `WorkerPool`, hence the need to run this operation here.

This function runs on the typechecker thread, _but_ the indexer thread is blocked on `startedNotification` so we will not race on the indexer resource here.

We use this pattern in one other place -- initialization. The commonality is a need to parallelize indexer operations.

WorkerPool _is_ threadsafe, so we could consider sharing it with the dispatch logic. I resist doing that because I do not want it lurked into places it does not belong, causing pipeline stalls because new edits are waiting for the slow path to finish using the WorkerPool. It's possible that a simple constrained API would be fine for this purpose.